src/lib/utils/common.ts

```diff
@@ -56,3 +56,20 @@ export const rankToString = (rank: number) => {
 export const toPercent = (decimal: number) => {
 	return `${(decimal * 100).toFixed(2)}%`;
 };
+
+/**
+ * Returns the first numDuplicates earliest duplicates in an array, in order
+ * If there aren't that many duplicates, return all duplicates (can be empty array)
+ */
+export const getDuplicates = <Type>(array: Type[]): Type[] => {
+	// Set.has is O(1), so use Sets
+	const seenElements = new Set<Type>();
+	const duplicates = new Set<Type>();
+	for (const elem of array) {
+		if (seenElements.has(elem)) {
+			duplicates.add(elem);
+		}
+		seenElements.add(elem);
+	}
+	return [...duplicates];
+};
```

Comment:
Typescript generic naming conventions + use readonly to allow use of readonly arrays (it won't force it) https://tsplay.dev/mpg2zN
```suggestion
export const getDuplicates = <T>(array: readonly T[]): T[] => {
	// Set.has is O(1), so use Sets
	const seenElements = new Set<T>();
	const duplicates = new Set<T>();
	for (const elem of array) {
		if (seenElements.has(elem)) {
			duplicates.add(elem);
		}
		seenElements.add(elem);
	}
	return [...duplicates];
};
```